### PR TITLE
Feature/detectioncrop element

### PR DESCRIPTION
**Pull request by tvlenin**

### CODE REVIEW

File: gst-libs/gst/r2inference/Makefile.am [hunk 1]
```diff
@@ -45,4 +45,10 @@ gstinferenceinclude_HEADERS=		\
 	gsttensorflow.h 		\
 	gstinferencepreprocess.h 	\
 	gstinferencepostprocess.h	\
-	gstinferencedebug.h
+	gstinferencedebug.h             \
+        gstinferencemeta.h
+
```
Comment (maumonteroj) [inline gst-libs/gst/r2inference/Makefile.am]: Remove extra blank lines

File: gst/inferencecrop/Makefile.am [hunk 2]
```diff
@@ -0,0 +1,37 @@
+plugin_LTLIBRARIES = libgstinferencecrop.la
+
+libgstinferencecrop_la_SOURCES =	\
+	gstdetectioncrop.cc             \
+        cropelement.cc                  \
+        videocrop.cc
+
```
Comment (maumonteroj) [inline gst/inferencecrop/Makefile.am]: Left only one blank line

File: gst/inferencecrop/videocrop.cc [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        }else{
+          left = left - round(((width_ratio * height) / height_ratio - width) / 2);
+          right = right - round(((width_ratio * height) / height_ratio - width) / 2);
+        }
+      }
+    }
+
+  }
+  
+
```
Comment (maumonteroj) [inline gst/inferencecrop/videocrop.cc]: Remove this blank line

File: gst/inferencecrop/videocrop.cc [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+void
+VideoCrop::UpdateElement (GstElement * element,
+			  gint image_width,
+			  gint image_height,
+			  gint x,
+			  gint y,
+			  gint width,
+			  gint height,
+        gint width_ratio,
+        gint height_ratio)
```
Comment (maumonteroj) [inline gst/inferencecrop/videocrop.cc]: Check this indentation

File: gst-libs/gst/r2inference/Makefile.am (hunk 1 above)
Comment (maumonteroj) [reply gst-libs/gst/r2inference/Makefile.am]: Check the indentation

File: gst/inferencecrop/gstdetectioncrop.cc [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
+static void
+gst_detection_crop_set_caps (GstPad * pad, GParamSpec * unused,
+    GstDetectionCrop * self)
+{
+  GstCaps *caps;
+  GstStructure *st;
+  gint width;
+  gint height;
+
+  g_object_get (pad, "caps", &caps, NULL);
```
Comment (michaelgruner) [inline gst/inferencecrop/gstdetectioncrop.cc]: Double check if this caps need to be freed.

File: gst/inferencecrop/gstdetectioncrop.cc [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
+  gint crop_height_ratio;
+  gint requested_index;
+  BBox box;
+  GstPadProbeReturn ret = GST_PAD_PROBE_DROP;
+
+  GST_OBJECT_LOCK (self);
+  self->aspect_ratio_tokens = g_strsplit (self->aspect_ratio, ":", 0);
+  crop_index = self->crop_index;
+  crop_class = self->crop_class;
+  crop_width_ratio = atoi(self->aspect_ratio_tokens[0]);
```
Comment (michaelgruner) [inline gst/inferencecrop/gstdetectioncrop.cc]: This is very dangerous. Where do you validate if the user entered a valid integer? as per now "A:1234" is a valid entry and will crash our plug-in. I recommend using a fraction instead.

File: gst/inferencecrop/videocrop.cc [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+{
+  gint top = y;
+  gint bottom = image_height - y - height;
+  gint left = x;
+  gint right = image_width - x - width;
+  if(width_ratio != 0 || height_ratio != 0 ){
```
Comment (michaelgruner) [inline gst/inferencecrop/videocrop.cc]: This file is not properly indented. Please use `gst-indent`

Comment (michaelgruner) [inline gst/inferencecrop/videocrop.cc]: What happens if one of them is zero?

File: gst/inferencecrop/videocrop.cc [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    }else{
+      if(width < height){
+        left = left - round(((width_ratio * height) / height_ratio - width) / 2);
+        right = right - round(((width_ratio * height) / height_ratio - width) / 2);
+      }else{
+        if(width > (width_ratio * height)){
+          top = top - round(((height_ratio * width) / width_ratio - height) / 2);
+          bottom = bottom - round(((height_ratio * width) / width_ratio - height) / 2);
+        }else{
+          left = left - round(((width_ratio * height) / height_ratio - width) / 2);
```
Comment (michaelgruner) [inline gst/inferencecrop/videocrop.cc]: This line is exactly the same as the one in line 64 and the line 58. Smells like refactoring.

File: gst/inferencecrop/videocrop.cc [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+{
+  gint top = y;
+  gint bottom = image_height - y - height;
+  gint left = x;
+  gint right = image_width - x - width;
+  if(width_ratio != 0 || height_ratio != 0 ){
+
+    if(width_ratio <= height_ratio){
+      if(width > height){
+        top = top - round(((height_ratio * width) / width_ratio - height) / 2);
```
Comment (michaelgruner) [inline gst/inferencecrop/videocrop.cc]: The second operand is the same for everyone. Factor out!

File: gst/inferencecrop/videocrop.cc [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+        gint width_ratio,
+        gint height_ratio)
+{
+  gint top = y;
+  gint bottom = image_height - y - height;
+  gint left = x;
+  gint right = image_width - x - width;
+  if(width_ratio != 0 || height_ratio != 0 ){
+
+    if(width_ratio <= height_ratio){
```
Comment (michaelgruner) [inline gst/inferencecrop/videocrop.cc]: This logic can be rewritten to avoid lots of duplicate code.

File: gst/inferencecrop/videocrop.h [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+class VideoCrop : public CropElement
+{
+public:
+  const std::string GetFactory () const override;
+  GstPad * GetSrcPad () override;
+  GstPad * GetSinkPad () override;
+
+protected:
+  void UpdateElement (GstElement * element,
+		      gint image_width,
```
Comment (michaelgruner) [inline gst/inferencecrop/videocrop.h]: Indent

File: gst/inferencecrop/gstdetectioncrop.cc [hunk 12]
```diff
@@ -0,0 +1,464 @@
...
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "gstdetectioncrop.h"
+
+#include "gst/r2inference/gstinferencemeta.h"
+#include "videocrop.h"
```
[the remark this answers is not in the document]
Comment (tvlenin) [reply gst/inferencecrop/gstdetectioncrop.cc]: We need this to share the default aspect-ratio

File: gst/inferencecrop/gstdetectioncrop.cc [hunk 13]
```diff
@@ -0,0 +1,473 @@
...
+
+#define PROP_CROP_INDEX_DEFAULT 0
+#define PROP_CROP_INDEX_MAX G_MAXUINT
+#define PROP_CROP_INDEX_MIN 0
+
+#define PROP_CROP_CLASS_DEFAULT -1
+#define PROP_CROP_CLASS_MAX G_MAXINT
+#define PROP_CROP_CLASS_MIN -1
+
+
```
Comment (maumonteroj) [inline gst/inferencecrop/gstdetectioncrop.cc]: Remove this blank lines

File: gst/inferencecrop/cropelement.h [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+  gint x;
+  gint y;
+  gint width;
+  gint height;
+  std::mutex mutex;
+  gint width_ratio;
+  gint height_ratio;
+};
+
+#endif //__CROP_ELEMENT_H__
```
Comment (maumonteroj) [inline gst/inferencecrop/cropelement.h]: Add a blank line at the final

File: gst/inferencecrop/videocrop.cc [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+{
+  gint top = y;
+  gint bottom = image_height - y - height;
+  gint left = x;
+  gint right = image_width - x - width;
+
+  if(width_ratio != 0){
+    gint top_bottom_modify = round(((height_ratio * width) / width_ratio - height) / 2);
+    gint left_right_modify = round(((width_ratio * height) / height_ratio - width) / 2);
+    if(width_ratio <= height_ratio){
```
Comment (michaelgruner) [inline gst/inferencecrop/videocrop.cc]: This file is poorly indented. Please use `gst-indent` in all files.

File: gst/inferencecrop/videocrop.cc [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+      if(width > height && height <= (height_ratio * width)){
```
Comment (michaelgruner) [inline gst/inferencecrop/videocrop.cc]: I don't understand the second condition. Why is it necessary?

File: gst/inferencecrop/videocrop.cc [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+                          gint height,
+                          gint width_ratio,
+                          gint height_ratio)
+{
+  gint top = y;
+  gint bottom = image_height - y - height;
+  gint left = x;
+  gint right = image_width - x - width;
+
+  if(width_ratio != 0){
```
[the remark this answers is not in the document]
Comment (tvlenin) [reply gst/inferencecrop/videocrop.cc]: The Fraction type allows a numerator = 0, but don't allow a denominator = 0, and through an erroneous pipeline warning, for that reason, I only check the numerator (width_ratio).

File: gst/inferencecrop/videocrop.cc (hunk 16 above)
Comment (tvlenin) [reply gst/inferencecrop/videocrop.cc]: I checked and it is no necessary.

File: gst/inferencecrop/cropelement.cc [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+{
+  GstElement * element;
+  bool validated = Validate();
+  this->mutex.lock ();
+  if(validated){
+    element = this->element;
+  }
+  this->mutex.unlock ();
+
+  return element;
```
[the remark this answers is not in the document]
Comment (tvlenin) [reply gst/inferencecrop/cropelement.cc]: No, I don't get a warning, but I add an evaluation to through an error if it is uninitialized.

**Pull request closed without merging**